### ISSUE TITLE
Handshake at startup

### ISSUE THREAD

**Issue by pat1shea** [jean-emmanuel/open-stage-control#341]
Is there a way for the app to send an initial message to synchronize with a remote app? I tried the init() from the custom module but it executes before the app is loaded so it can't get initial values

**Comment by jean-emmanuel** [comment]
I'm thinking of exposing an event emitter to the custom module's scope, this way you could bind callbacks to events such as "session-opened" for example. Something like:
```js
init: function(){
    app.on('session-opened', function(clientId, sessionName) {
        // ...
    })
}

**Comment by pat1shea** [comment]
Yes that would be great. The purpose of this is to re-syn the app on restart with the actual controller app so that it can reflect the current state.
 
Also it would be nice if the controls had a disabled property.
 
Thanks
Patrick
-----Original Message-----

**Comment by cyberic99** [comment]
The ``session-opened`` callback would be useful for me as well.

For now I can detect o-s-c with zeroconf, but it annonces itself before opening the session, so it is too early and I don't have the session name

Another way of doing it could be to send an OSC event like /SESSION_LOADED <session name> to every synced target

**Comment by jean-emmanuel** [comment]
>Also it would be nice if the controls had a disabled property.
check this out https://osc.ammd.net/extras/css-tips/#layering-pointer-events

**Comment by jean-emmanuel** [comment]
I just push a very simple implementation that exposes `app` to the custom module's scope: it is an [event emitter](https://nodejs.org/api/events.html#events_class_eventemitter) for monitoring the events sent by the different clients. Event names can be found in [callbacks.js](https://github.com/jean-emmanuel/open-stage-control/blob/master/app/main/callbacks.js), callbacks are called with 2 arguments: `data` (object) and `clientId` (string, a unique id that identify each client browser session).

```js
init: function(){
    app.on('sessionOpen', function(data, id) {
        console.log(`Client ${id} opened session file ${data.path}`)
    })
}
```

**Comment by pat1shea** [comment]
What's the best way to integrate changes back into the installed app? Just copy the sources from app to resources/app ?
 
Patrick
-----Original Message-----

**Comment by jean-emmanuel** [comment]
The best way is to follow the [run/build from souces](https://osc.ammd.net/getting-started/#run-from-sources) instructions, or to wait for the next release, which is beeing uploaded right now.

**Comment by jean-emmanuel** [comment]
I'm closing the issue, feel free to reopen if the features needs improvements

**Comment by pat1shea** [comment]
This doesn't seem to work, looks like the event is not propagated to "app", is "app" the same as "ipc"?

Is there a catch all event to test if any events are captured by "app" ?

**Comment by jean-emmanuel** [comment]
Attached custom module file works here: [test.js.txt](https://github.com/jean-emmanuel/open-stage-control/files/2231440/test.js.txt)

**Comment by pat1shea** [comment]
Okay I see now, this is a browser session. What I was looking for is an event to broadcast to targets (--send list) to notify when the server is up. In my case this is a java app that is already running and need to re-sync the state values to the osc app.
 
The issue was that in the init() method the server is not open yet so it cannot accept messages from osc client.
 
Ideally it would be a callback with the list of targets so they can be notified.
 
Patrick
 
-----Original Message-----

**Comment by jean-emmanuel** [comment]
You can retreive the `--send` value with `settings.read('targets')`, it was missing from the docs until now.

**Comment by pat1shea** [comment]
Thanks
 
On v34 I get the following error that I don't get on v33
 
TypeError: Cannot read property 'send' of undefined
at OscServer.receiveOsc (C:\Users\DPSH510\apps\open-stage-control-0.34.0-win32-x64\resources\app\main\osc\index.js:140:12)
at OscServer.oscInHandler (C:\Users\DPSH510\apps\open-stage-control-0.34.0-win32-x64\resources\app\main\osc\index.js:165:14)
at OscServer.oscInHandlerWrapper (C:\Users\DPSH510\apps\open-stage-control-0.34.0-win32-x64\resources\app\main\osc\index.js:175:18)
at emitThree (events.js:135:13)
at osc.UDPPort.emit (events.js:216:7)
at Object.osc.firePacketEvents (C:\Users\DPSH510\apps\open-stage-control-0.34.0-win32-x64\resources\app\node_modules\osc\src\osc-transports.js:23:18)
at osc.UDPPort.p.decodeOSC (C:\Users\DPSH510\apps\open-stage-control-0.34.0-win32-x64\resources\app\node_modules\osc\src\osc-transports.js:83:17)
at emitTwo (events.js:125:13)
at osc.UDPPort.emit (events.js:213:7)
at Socket.
-----Original Message-----

**Comment by jean-emmanuel** [comment]
Damned, hot fix on the way...

**Comment by jean-emmanuel** [comment]
patched v0.34.1 is being uploaded, thanks.

**Comment by pat1shea** [comment]
I still had issues with this Jean-Emmanuel but here's what I did to make this work (I'm a veteran java developer but new to nodejs so this may not be ok!)
 
I exposed customModule in index.js like this:
 
module.exports = {
customModule: oscServer.customModule,
...
 
Then in callbacks.js sessionOpen(...) I added this
 
if (osc.customModule && osc.customModule.sessionOpen) {
 osc.customModule.sessionOpen(data,clientId)
 }
 
 
Patrick
-----Original Message-----

**Comment by pat1shea** [comment]
Finally I changed it to this
 
index.js ->
module.exports = {
sessionOpen: function(data,clientId) {
 if (oscServer.customModule.sessionOpen) {
 this.customModule.sessionOpen(data,clientId)
 }
 },
 
callbacks.js ->
osc.sessionOpen(data,clientId)
 
 
-----Original Message-----

**Comment by jean-emmanuel** [comment]
The evented `app` object already provide a way to do this, could provide a custom module that doesn't work as expected ?

**Comment by pat1shea** [comment]
Version 34.1 does not work at all, the browser window comes up blank, even for examples
 
I find it a little weird to use events when each object already contains a reference to each other. Normally I use events on things that are not directly related.
 
Patrick
-----Original Message-----

**Comment by jean-emmanuel** [comment]
> Version 34.1 does not work at all, the browser window comes up blank, even for examples

Can you try with the latest prebuilt binaries (v0.34.2)?

**Comment by pat1shea** [comment]
same result
-----Original Message-----

**Comment by pat1shea** [comment]
Okay it seems that the issue is with the port parameter. If blank or 8080 it works, anything else does not work.
 
Patrick
 
-----Original Message-----

**Comment by pat1shea** [comment]
The issue is here:
 
settings.js line 154 is:
 
appAddresses.push('http://' + ifaces[ifname][i].address + ':' + (argv.p || 8080))
 
and should be:
 
appAddresses.push('http://' + ifaces[ifname][i].address + ':' + (argv['port'] || 8080))
 
-----Original Message-----

**Comment by jean-emmanuel** [comment]
Yay, thank you ! Fixed in 0.34.3
